File: typescript/ch5/IO.ts
import _ from 'lodash';
import { WrapperFunc } from './WrapperFunc';

type IOFunc = () => any;

class IO {
  protected _effect: IOFunc;

  constructor(effect: IOFunc) {
    if (!_.isFunction(effect)) {
      throw 'IO 사용법: 함수는 필수입니다!';
    }
    this._effect = effect;
  }

  static of(value: any) {
    return new IO(() => value);
  }

  static from(f: IOFunc) {
    return new IO(f);
  }

  map(f: WrapperFunc) {
    return new IO(() => f(this._effect()));
  }

  chain(f: WrapperFunc) {
    return f(this._effect());
  }

  run() {
    return this._effect();
  }
}

export {
  IO
};

File: python/python/ch5/IO.py
import pydash

class IO:
    def __init__(self, effect):
        if not pydash.is_function(effect):
            raise TypeError('IO 사용법: 함수는 필수입니다!')
        self._effect = effect
        
    @staticmethod
    def of(value):
        return IO(lambda: value)
      
    @staticmethod
    def from_func(f):
        return IO(f)
      
    def map(self, f):
        return IO(lambda: f(self._effect()))
      
    def chain(self, f):
        return f(self._effect())
      
    def run(self):
        return self._effect()